Recently used and favorite payees

qedi-r opened the pull request:
<!-- Thank you for submitting a pull request! Make sure to follow the instructions to write release notes for your PR — it should only take a minute or two: https://github.com/actualbudget/docs#writing-good-release-notes -->

I wanted to improve on the alphabetized payees I always saw by default when the payee dropdown showed up. It never looked helpful, just the numbered companies I've used once in the past. Instead, I got the list of the 10 most frequently used payees in the past 3 months, and used that first, then the rest of the payees, then the transfer section. Typing still works to find what you need.

![image](https://github.com/actualbudget/actual/assets/1435081/f9ae8863-87fb-432b-86a8-6dcad59a8d65)

Then I decided being able to mark some that might not be always frequently used but useful to show in the list would be good, so I also added a 'favorite' for payees. The screenshot shows the 10 most frequently used payees as stars, things marked as favorite with the favorite icon, and others as the empty favorite icon. You can click on them to mark them as favorite or un-favorite them. 

![image](https://github.com/actualbudget/actual/assets/1435081/51a16b05-40de-4c97-8d0f-9cceccbaabc6)

"Recently used payees" shows both the top 10 most used (by number of transactions in the past 3 months) plus favorites.

One more note, I thought finding the recently used payees could end up being an expensive operation, based on the sql, so I didn't throw it in PayeeEntity by default, you must query for it specifically, to avoid it being calculated where it's not needed.

TODO:
- [x] release notes
- [x] fixing tests
- [x] ensuring additional test coverage exists

qedi-r commented:
I think this is ready for review now - I would have liked to add more tests, though I'm not familiar enough with the testing infra yet to understand what's possible, specifically around the new query to guard against regression.

youngcw commented:
A few thoughts

* I think the title in the autocomplete should be shortened to just "Frequent" or "Recently Used".  That would match the other headers length better.
* If I have a payee marked as favorite, then it shows up in a different section of the autocomplete _after_ the bulk payee section.  Shouldn't that be at the top or in the frequently used list?
* Is it that useful to mark the frequent payees in the payee list?  Cant those just populate on their own without being called out in the table?
* Continuing on the previous, I don't really like the current empty bookmark for a non-favorite payee.  Between the stars and empty bookmarks it makes it look like all the payees are already marked instead of the favorites having a way to stand out.  Maybe instead of a bookmark icon we could have a bottom appear for selected payees?  More in line with merging and deleting?

Teprifer commented:
* I think I echo youngcw's comments, particularly around merging the bookmarked in to the frequent section with one addition which would be to stop adding frequently used payees if the list size would go over 10. My thinking here is to avoid the list getting too long, while allowing people to override by bookmarking as many as they like. 

   e.g. 3 bookmarked payees would mean 7 frequently used were added, but 12 bookmarked would mean all 12 would show, and no frequently used were added

* Ditto on no need to indicate frequent payees in the list and dialing back the indicator of not bookmarked.

* One thought, and don't feel obligated to implement this, is a way to exclude payees from the frequently used list.
This thought came to mind as my frequently used payee list includes my bank for mortgage payments as \# 2 . Likely down to the fact I pay weekly, and two different payments each time, so it'll always be frequent. But in reality I'll almost never add that payee manually as schedules take care of adding those transactions.
Others may want to similarly exclude payees that are only added/used by schedules.

* Consider adding 'Add as bookmark' / 'Remove bookmark' (or whatever text makes sense) options to the payee options menu to increase discoverability(also easier to click)? Likewise 'Exclude from Frequent Payees' could be added here if the above suggestion is implemented.
![image](https://github.com/actualbudget/actual/assets/34882928/da36cc7c-3d1e-45be-92ed-0e81fe55c95f)

* Bug: the payee list seems to get a bit discombobulated and adding additional sections for frequently used payees and payees if you add a couple as bookmarked, navigate away, come back and bookmark some more.

![image](https://github.com/actualbudget/actual/assets/34882928/0819b038-47b4-46d1-a32e-d27fd5eb7fb9)
_To see this in the demo budget well you'll need to create some additional payees by adding some additional transactions_
e.g. 
![image](https://github.com/actualbudget/actual/assets/34882928/3ec9be74-e183-4927-9e1a-a154b36ffd67)

qedi-r commented:
Yeah, I think these are reasonable suggestions. I've fixed the resorting and bugs, and cleaned up the bookmarking into the menu. I've removed the ability to click on the icons, just the menu seems more like the way to do it. It feels like there is a pattern of just applying the flag to all selected items if there are some items without the flag, and removing it if they are all flagged. This is how clearing works for transactions, so this is how I implemented bookmarking.

qedi-r commented:
I also named it "Recently Used" and maxed it out to 10, as [Teprifer](https://github.com/Teprifer) suggested.

youngcw commented:
If a frequently used payee is the same as a favorited payee, that payee gets listed twice top list

psybers commented:
I'm sure it is out of scope of this specific PR, but the list of 'frequent' payees, in my opinion, should probably be the frequent ones you *manually* enter, and not ones set automatically by rules or ones that are just the most frequent in the recent X months.

psybers commented:
> If a frequently used payee is the same as a favorited payee, that payee gets listed twice.

And in such a case, you get 9 unique payees listed.  So it actually lost one.

psybers commented:
Personal opinion but I'd rather just see 3-5 frequent payees listed.  Much more than that and odds are I am better off just typing and searching for what I want.

qedi-r commented:
I didn't think the common payees bit would be all that divisive - should I split it out of this, keep the favourites, and create a feature request for it to get a more general set of opinions? Five might be a better number, too.

qedi-r commented:
Second question - is this intentional?
![image](https://github.com/actualbudget/actual/assets/1435081/8a1317d5-0067-46b4-bce8-b021bc902660)

Only on the all accounts, if you open Payees only transfers show up as options. I assumed this was intentional, but once I found the code that was supposed to do this, TransactionsTable.tsx:465-457, I don't think it is. You can end up accidentally creating a new payee with the same name as an existing one if you just type the name and move on, as it won't be selected in this case.

psybers commented:
> Second question - is this intentional? ![image](https://private-user-images.githubusercontent.com/1435081/339999550-8a1317d5-0067-46b4-bce8-b021bc902660.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg0NzkyMzQsIm5iZiI6MTcxODQ3ODkzNCwicGF0aCI6Ii8xNDM1MDgxLzMzOTk5OTU1MC04YTEzMTdkNS0wMDY3LTQ2YjQtYmNlOC1iMDIxYmM5MDI2NjAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYxNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MTVUMTkxNTM0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MjBmZGViYThjZGJhYmFhZDY5NTdlZmY5OGFiNTA2MjViZDZmMGQ4NjJhZTcxMDYyNmJmYmY0Mjc0N2M3MGY5NCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.sWcY40OwLnCpHcM_ar0KAH1f2a6NdyiJ1z0l2Kkwi7E)
> 
> Only on the all accounts, if you open Payees only transfers show up as options. I assumed this was intentional, but once I found the code that was supposed to do this, TransactionsTable.tsx:465-457, I don't think it is. You can end up accidentally creating a new payee with the same name as an existing one if you just type the name and move on, as it won't be selected in this case.

This looks like maybe a bug, but it exists outside your PR so don't worry about it.  For what its worth if you pick a bank first, then the payees list.

shall0pass commented:
Have you considered time-boxing the favorites list? For example, if today is the 15th of the month, fill the favorites list based on previous months for the 12th thru 18th.  If anyone else is like me, I'm a creature of habit so I do a lot of the same things the same time each month.

 I'm not sure how well it would actually work in practice or which time range would work best.

I'm looking forward to the inclusion of this feature regardless.

qedi-r commented:
I'll file the bug and fix it separately then.

While I like the idea of payee commonly used at particular times,  I feel like that would be a complex query,  or require data we don't currently have.

youngcw commented:
@qedi-r I think the only thing that still needs fixed is https://github.com/actualbudget/actual/pull/2814#issuecomment-2165916122

We probably can get this merged after the release if that gets fixed

qedi-r commented:
Ah, thanks for commenting, I was starting to wonder if I didn't follow procedure at the end. I did fix that in my last set of commits. (FYI, I had to force-reload the demo preview, it had cached older versions of the PR UI)

qedi-r commented:
A question though - when looking at it today to respond, I got an error "Database is out of sync with migrations". Is that expected because there was a version bump while the PR was opened, and it was expected, or is there something I didn't do correctly for the migration?

youngcw commented on packages/loot-core/src/server/api.ts:
```diff
@@ -606,6 +606,12 @@ handlers['api/category-delete'] = withMutation(async function ({
   });
 });
 
+handlers['api/payees-get'] = async function () {
```
Should this be `common-payees-get`?

youngcw commented on upcoming-release-notes/2814.md:
```diff
@@ -0,0 +1,6 @@
+authors: [qedi-r]
+---
+
+Shows favourite and up to the top 10 most frequently used payees in the payee dropdown menu in a section at the top.
```
the number has changed

qedi-r replied:
Thanks!

youngcw commented:
> A question though - when looking at it today to respond, I got an error "Database is out of sync with migrations". Is that expected because there was a version bump while the PR was opened, and it was expected, or is there something I didn't do correctly for the migration?

your file must have come from a build with more migrations.  I always test using incognito and a demo file unless I explicitly need to test a pre existing file

youngcw commented:
there have been a few migrations since you created this PR so the number on the migration will need upped to be in order with the existing ones.

youngcw commented:
Duplication seems to be happening only with the "Deposit" payee.  I favorited all the others in this list and they are working.
![image](https://github.com/actualbudget/actual/assets/28542559/a0195663-76c6-4b4b-8567-c5c7e5555d59)

youngcw commented:
Ok, this is another nitpik.  The section in the payee list should probably include something to call out that there may be favorites in the list. Like "Recent\Favorites" or similar.  I know I said just "frequent" before, but that guy was silly.

psybers commented:
> Ok, this is another nitpik. The section in the payee list should probably include something to call out that there may be favorites in the list. Like "Recent\Favorites" or similar. I know I said just "frequent" before, but that guy was silly.

Maybe favorited ones should have a star next to them?

youngcw commented on packages/loot-core/src/server/aql/schema/index.ts:
```diff
@@ -61,6 +61,7 @@ export const schema = {
     name: f('string', { required: true }),
     transfer_acct: f('id', { ref: 'accounts' }),
     tombstone: f('boolean'),
+    favorite: f('boolean'),
```
:clap: **Praise**: Thanks for remembering this part.  It usually gets forgotten.

qedi-r commented:
Ok, I simplified it to "Suggested Payees" but added the same bookmark icon on the payee list to give clarity where it's coming from. And I fixed the bug of the single payee duplicating. Seemed to be a caching issue between the two queries' responses.

![image](https://github.com/user-attachments/assets/e6bde595-6fb1-4e06-b5c5-d317fd6184ec)

youngcw commented on packages/loot-core/migrations/1720664867241_add_payee_favorite.sql:
```diff
@@ -0,0 +1,5 @@
+BEGIN TRANSACTION;
+
+ALTER TABLE payees ADD COLUMN favorite INTEGER DEFAULT 0 DEFAULT FALSE;
```
Are there two defaults on purpose?  Its working so its not that big of a deal I don't think.
